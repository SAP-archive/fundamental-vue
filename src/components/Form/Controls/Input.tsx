import {
  Watch,
  Inject,
} from 'vue-property-decorator';
import { Component, Event, Prop, Base } from '@/core';
import { FORM_ITEM_ID_KEY } from './../FormItem';

interface Props {
  id?: string | null;
  placeholder?: string;
  required?: boolean;
  disabled?: boolean;
  state?: InputState | null;
  type?: InputType;
  readonly?: boolean;
  value?: string | number | null;
  compact?: boolean;
  maxlength?: string | null;
}
import { isInputElement } from './Helper';

const typeMappings = {
  text: 'Text Field',
  password: 'Password Field',
  search: 'Search Field',
};
type InputType = keyof (typeof typeMappings);
const InputTypes = Object.keys(typeMappings) as InputType[];

const stateMapping = {
  valid: 'Valid',
  invalid: 'Invalid',
  warning: 'Warning',
};
type InputState = keyof (typeof stateMapping);
const InputStates = Object.keys(stateMapping) as InputState[];

@Component('Input')
@Event('input', 'Sent when the value changes', ['value', 'any'])
@Event('change', 'Sent when the value finishes to change', ['value', 'any'])
export class Input extends Base<Props> {

  @Inject({from: FORM_ITEM_ID_KEY, default: null})
  public formItemId!: string | null;

  @Prop('placeholder text', { default: '', type: String })
  public placeholder!: string;

  @Prop('current state', { acceptableValues: InputStates, default: null, type: String })
  public state!: InputState | null;

  @Prop('whether a value is required (adds a *)', { default: false, type: Boolean })
  public required!: boolean;

  @Prop('button type', { acceptableValues: InputTypes,  default: 'text', type: String })
  public type!: InputType;

  @Prop('whether the control is disabled', { default: false, type: Boolean })
  public disabled!: boolean;

  @Prop('whether the control is readonly', { default: false, type: Boolean })
  public readonly!: boolean;

  @Prop('current value', { default: '', type: [String, Number] })
  public value!: string | null;

  @Prop('whether input is compact', { type: Boolean, default: false })
  public compact!: boolean;

  @Prop('upper limit on the number of characters that can be entered in the input field.', {type:String})
  public maxlength!: string | null;

  private get inputId(): string | null {
    return this.formItemId;
  }

  public render() {
    return (
        <input
          id={this.inputId}
          value={this.currentValue}
          on-input={this.updateInput}
          on-change={this.hasChange}
          readonly={this.readonly}
          disabled={this.disabled}
          type={this.type}
          class={this.classes}
          staticClass='fd-form__control'
          placeholder={this.placeholder.length > 0 ? this.placeholder : false}
          maxlength={this.maxlength}
        />
    );
  }

  @Watch('value', { immediate: true })
  public handleNewValue(newValue: string | null) {
    this.currentValue = newValue;
  }

  private updateInput({ target }: Event) {
    if(target == null) {
      return;
    }
    if(!isInputElement(target)) {
      return;
    }
    const { value } = target;
    this.$emit('input', value);
  }
  private hasChange({target}: Event) {
    if(target == null) {
      return;
    }
    if(!isInputElement(target)) {
      return;
    }
    const { value } = target;
    this.$emit('change', value);
  }
  public currentValue: string | null = (this.value === undefined || this.value === null) ? '' : this.value;

  private get classes() {
    return {
      'fd-input--compact': this.compact,
      'is-warning': this.state === 'warning',
      'is-invalid': this.state === 'invalid',
      'is-valid': this.state === 'valid',
      'is-required': this.required,
    };
  }
}
