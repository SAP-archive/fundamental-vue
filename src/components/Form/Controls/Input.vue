<template>
  <textarea
    v-if="type === 'textarea'"
    :class="inputClasses"
    :id="inputId"
    :readonly="readonly ? '' : null"
    :disabled="disabled ? '' : null"
    :placeholder="placeholder"
    :value="value"
    @input="$emit('update', $event.target.value)"
    @change="$emit('change', $event.target.value)"
    v-on="$listeners"
    v-bind="$attrs"
  />
  <input
    v-else
    class="fd-input fd-form__control"
    :class="inputClasses"
    :id="inputId"
    :readonly="readonly ? '' : null"
    :disabled="disabled ? '' : null"
    :type="type"
    :placeholder="placeholder"
    :value="value"
    @blur="handleBlur"
    @focus="handleFocus"
    @input="$emit('update', $event.target.value)"
    @change="$emit('change', $event.target.value)"
    v-on="$listeners"
    v-bind="$attrs"
  />
</template>

<script>
import InputMixin from "./InputMixin";
import { $default, $valueWithDefault } from "./Helper/prop";

export default {
  methods: {
    handleBlur(event) {
      this.focused = false;
      this.$emit("blur", event);
    },
    handleFocus(event) {
      this.focused = true;
      this.$emit("focus", event);
    }
  },
  name: "FdInput",
  mixins: [InputMixin],
  model: {
    event: "update"
  },
  data() {
    return {
      focused: false
    };
  },
  props: {
    placeholder: $default(""),
    type: $default("text"),
    value: $valueWithDefault("")
  }
};
</script>
