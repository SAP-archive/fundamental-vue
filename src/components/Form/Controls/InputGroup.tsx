import { Slot, Component, DefaultSlot, Prop, Base } from '@/core';

interface Props {
  before?: string | null;
  after?: string | null;
  afterClass?: string | null;
  compact?: boolean;
}

@Component('InputGroup')
@Slot('before', 'Content to be placed before the input component.')
@Slot('after', 'Content to be placed after the input component.')
@DefaultSlot('The input component placed in the input group.')
export class InputGroup extends Base<Props> {
  @Prop('text/number before the input', { type: String, default: null })
  public before!: string | null;

  @Prop('text/number after the input', { type: String, default: null })
  public after!: string | null;

  @Prop({ type: String, default: null })
  public afterClass!: string | null;

  @Prop('whether input group is compact', { type: Boolean, default: false })
  public compact!: boolean;

  public render() {
    const beforeAddon = this.before || this.$slots.before;
    const afterAddon = this.after || this.$slots.after;
    const afterClass = this.afterClass || '';
    const ws = this.afterClass ? ' ' : '';
    const afterAddonClassName = 'fd-input-group__addon fd-input-group__addon--after' + ws + afterClass;
    return (
      <div staticClass='fd-input-group' class={this.classes}>
        {beforeAddon && <span class='fd-input-group__addon fd-input-group__addon--before'>{beforeAddon}</span>}
        {this.$slots.default}
        {afterAddon && <span class={afterAddonClassName}>{afterAddon}</span>}
      </div>
    );
  }

  private get classes() {
    return {
      'fd-input-group--before': this.before || this.$slots.before,
      'fd-input-group--after': this.after || this.$slots.after,
      'fd-input-group--compact': this.compact,
    };
  }
}
