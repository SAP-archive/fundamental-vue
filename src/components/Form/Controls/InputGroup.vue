<template>
  <div class="fd-input-group" :class="classes">
    <span
      v-if="hasBefore"
      class="fd-input-group__addon fd-input-group__addon--before"
    >
      <slot name="before">{{ before }}</slot>
    </span>
    <slot />
    <span v-if="hasAfter" :class="afterClasses">
      <slot name="after">{{ after }}</slot>
    </span>
  </div>
</template>

<script>
export default {
  name: "FdInputGroup",
  props: {
    before: { type: String, default: null },
    after: { type: String, default: null },
    afterClass: { type: String, default: null },
    compact: { type: Boolean, default: false }
  },
  computed: {
    hasBefore() {
      return this.before != null || this.$slots.before != null;
    },
    hasAfter() {
      return this.after != null || this.$slots.after != null;
    },
    afterClasses() {
      const afterClass = this.afterClass;
      return [
        "fd-input-group__addon",
        "fd-input-group__addon--after",
        this.compact ? "fd-input-group--compact" : "",
        ...(afterClass != null ? [afterClass] : [])
      ];
    },
    classes() {
      return {
        "fd-input-group--before": this.before || this.$slots.before,
        "fd-input-group--after": this.after || this.$slots.after,
        "fd-input-group--compact": this.compact
      };
    }
  }
};
</script>
