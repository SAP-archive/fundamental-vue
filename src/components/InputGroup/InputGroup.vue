<template>
  <div class="fd-input-group" :class="classes">
    <fd-addon-before-provider>
      <slot name="before">
        {{ before }}
      </slot>
    </fd-addon-before-provider>
    <slot name="before-input" />
    <slot name="input" />
    <slot name="after-input" />
    <fd-addon-after-provider>
      <slot name="after">{{ after }}</slot>
    </fd-addon-after-provider>
  </div>
</template>

<script>
import { CompactableContainer } from "./../../mixins";

const CreateAddon = context => ({
  provide() {
    return {
      $_fdInputGroupAddonContext: context
    };
  },
  render() {
    return this.$scopedSlots.default();
  }
});

const FdAddonBeforeProvider = CreateAddon("before");
const FdAddonAfterProvider = CreateAddon("after");

export default {
  name: "FdInputGroup",
  mixins: [CompactableContainer],
  components: { FdAddonBeforeProvider, FdAddonAfterProvider },
  props: {
    before: { type: String, default: null },
    after: { type: String, default: null }
  },
  computed: {
    hasBefore() {
      return this.before != null || this.$slots.before != null;
    },
    hasAfter() {
      return this.after != null || this.$slots.after != null;
    },
    afterClasses() {
      const afterClass = this.afterClass;
      return [
        "fd-input-group__addon",
        "fd-input-group__addon--after",
        this.finalCompact ? "fd-input-group--compact" : "",
        ...(afterClass != null ? [afterClass] : [])
      ];
    },
    classes() {
      return {
        "fd-input-group--before": this.before || this.$slots.before,
        "fd-input-group--after": this.after || this.$slots.after,
        "fd-input-group--compact": this.compact
      };
    }
  }
};
</script>
