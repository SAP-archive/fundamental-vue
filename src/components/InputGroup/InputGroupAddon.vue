<template>
  <span class="fd-input-group__addon" :class="classes"><slot /></span>
</template>

<script>
export default {
  name: "FdInputGroupAddon",
  inject: {
    $_fdInputGroupAddonContext: { default: null }
  },
  provide() {
    return {
      $_fdInputGroupAddon: this
    };
  },
  data() {
    return {
      hasButton: false
    };
  },
  computed: {
    addonContext() {
      return this.$_fdInputGroupAddonContext;
    },
    classes() {
      return {
        "fd-input-group__addon--button": this.hasButton,
        "fd-input-group__addon--before": this.addonContext === "before",
        "fd-input-group__addon--after": this.addonContext === "after"
      };
    }
  }
};
</script>
