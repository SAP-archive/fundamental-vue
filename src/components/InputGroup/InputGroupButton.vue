<template>
  <fd-button styling="light" v-bind="$attrs" v-on="$listeners"><slot /></fd-button>
</template>

<script>
import FdButton from "./../Button/Button.vue";

export default {
  name: "FdInputGroupButton",
  components: { FdButton },
  inject: ["$_fdInputGroupAddon"],
  mounted() {
    this.$_fdInputGroupAddon.hasButton = true;
  }
};
</script>
