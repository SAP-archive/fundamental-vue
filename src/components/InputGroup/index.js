import InputGroup from "./InputGroup.vue";
import InputGroupSearch from "./InputGroupSearch.vue";
import InputGroupAddon from "./InputGroupAddon.vue";
import InputGroupButton from "./InputGroupButton.vue";
import InputGroupClearButton from "./InputGroupClearButton.vue";
import { pluginify } from "./../../util";
export default pluginify(
  InputGroup,
  InputGroupAddon,
  InputGroupButton,
  InputGroupClearButton,
  InputGroupSearch
);
export {
  InputGroup,
  InputGroupAddon,
  InputGroupButton,
  InputGroupClearButton,
  InputGroupSearch
};
