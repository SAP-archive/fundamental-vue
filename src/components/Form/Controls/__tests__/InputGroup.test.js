import { mount } from "@vue/test-utils";
import InputGroup from "../InputGroup.vue";
import Input from "../Input.vue";

describe("InputGroup", () => {
  const before = "$";
  const after = "â‚¬";

  const wrapper = mount(InputGroup, {
    propsData: {
      before,
      after
    },
    slots: {
      default: {
        render(h) {
          return h(Input);
        }
      }
    }
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders before after correctly", () => {
    expect(wrapper.contains(".fd-input-group__addon--before")).toBe(true);
    expect(wrapper.find(".fd-input-group__addon--before").text()).toEqual(
      before
    );
    expect(wrapper.contains(".fd-input-group__addon--after")).toBe(true);
    expect(wrapper.find(".fd-input-group__addon--after").text()).toEqual(after);
  });

  describe("Compact Style", () => {
    const wrapper = mount(InputGroup, {
      propsData: {
        compact: true
      },
      slots: {
        default: {
          render(h) {
            return h(Input);
          }
        }
      }
    });

    it("renders correctly", () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it("has compact style", () => {
      expect(wrapper.contains(".fd-input-group--compact")).toBe(true);
    });

    it("does not have before and after", () => {
      expect(wrapper.contains(".fd-input-group__addon--before")).toBe(false);
      expect(wrapper.contains(".fd-input-group__addon--after")).toBe(false);
    });
  });
});
