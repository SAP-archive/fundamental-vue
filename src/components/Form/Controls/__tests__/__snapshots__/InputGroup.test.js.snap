// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputGroup Compact Style renders correctly 1`] = `
<div
  class="fd-input-group fd-input-group--compact"
>
  <!---->
   
  <input
    class="fd-input fd-form__control"
    placeholder=""
    type="text"
  />
   
  <!---->
</div>
`;

exports[`InputGroup renders correctly 1`] = `
<div
  class="fd-input-group fd-input-group--before fd-input-group--after"
>
  <span
    class="fd-input-group__addon fd-input-group__addon--before"
  >
    $
  </span>
   
  <input
    class="fd-input fd-form__control"
    placeholder=""
    type="text"
  />
   
  <span
    class="fd-input-group__addon fd-input-group__addon--after"
  >
    â‚¬
  </span>
</div>
`;
